util/fpga/splice.py

```diff
@@ -0,0 +1,381 @@
+
+            if not infile.endswith('.bit'):
+                os.symlink(infile, f'{tmp}/infile.bit')
+                infile = f'{tmp}/infile.bit'
+
+            # Splice each data file into the bitstream.
+            for i, mmi in enumerate(mmis):
+                tmpbit = f'{tmp}/{i}.bit'
+                self.updatemem(infile, mmi, f'{tmp}/{i}.mem', tmpbit)
+                infile = tmpbit
+
+            # Update the usr_access_value to make this bitstream uniquely identifiable.
+            self.update_usr_access(infile, f'{tmp}/updated.bit')
+
+            # Copy the final bitstream into the cache and return the bitstream content.
+            with open(f'{tmp}/updated.bit', 'rb') as f:
+                content = f.read()
+                if outfile:
+                    outfile.seek(0)
+                    outfile.truncate()
+                    outfile.write(content)
+                    outfile.seek(0)
+                return content
+
+
+def main(args):
+    mmi = args.mmi.split(',')
+    data = args.data.split(',')
+    if len(data) == 0:
+        logging.error('No data files to splice')
+        return 1
+    if len(data) != len(mmi):
+        logging.error(
+            'There must be a 1:1 correspondence between data and mmi files')
+        return 1
+
+    cachedir = os.path.expanduser(args.cache)
+    os.makedirs(os.path.join(cachedir, 'splice'), exist_ok=True)
+
+    tools = Tools(args)
+    key = tools.hash([args.bitstream] + data + mmi)
+
+    # We open and lock the cache's bitstream file so that any simultaneous
+    # attempts to create or access the bitstream file will block here
+    # and allow one process to generate the bitstream.
+    filename = f'splice/{key}.bit'
+    bitstream = open_locked(os.path.join(cachedir, filename), binary=True)
+
+    logging.info(f'Checking cache for bitstream {key}')
+    # We open and lock the cache control file so that access or updates of
+    # the control file are serialized.
+    cache = CacheControlFile(f'{cachedir}/splices.json')
+
+    # Expire old bitstreams to prevent the cache from getting too big.
+    expired = cache.expire(args.cache_entries)
+    tools.remove_files(
+        os.path.join(cachedir, ex['filename']) for ex in expired.values())
+
+    splice = cache.get(key)
+    if splice is None:
+        logging.info('Bitstream doesn\'t exist.  Creating it.')
+        splice = cache.create(key, filename)
+    elif filename != splice['filename']:
+        logging.warning('Bitstream file has moved.  Updating.')
+        splice = cache.create(key, filename)
+    else:
+        logging.info(f'Bitstream is {splice["filename"]}')
+
+    content = bitstream.read()
+    digest = hashlib.sha256(content).hexdigest()
+
+    if digest == splice['digest']:
+        # If we didn't have to generate the bitstream, just update the access
+        # time and close the cache.
+        cache.access(key)
+        cache.close()
```

Comment:
nit: Would it make sense to consider using `contextmanager` or similar for all objects handling file descriptors? This way all operations could be wrapped with `with` statements.